src/constraint.rs

```diff
@@ -703,3 +715,57 @@ impl ConstraintValidator for FieldsConstraint {
         Ok(())
     }
 }
+
+/// Implements Ion Schema's `contains` constraint
+/// [contains]: https://amzn.github.io/ion-schema/docs/spec.html#contains
+#[derive(Debug, Clone, PartialEq)]
+pub struct ContainsConstraint {
+    // TODO: convert this into a HashSet once we have an implementation of Hash for OwnedElement in ion-rust
+    // Reference ion-rust issue: https://github.com/amzn/ion-rust/issues/220
+    values: Vec<OwnedElement>,
+}
+
+impl ContainsConstraint {
+    pub fn new(values: Vec<OwnedElement>) -> Self {
+        Self { values }
+    }
+}
+
+impl ConstraintValidator for ContainsConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // get the expected values from the contains constraint
+        let mut expected_values = self.values.iter();
+
+        // Check for null sequence
+        if value.is_null() {
+            return Err(Violation::new(
+                "contains",
+                ViolationCode::TypeMismatched,
+                &format!("Null sequence not allowed for contains constraint"),
+            ));
+        }
+
+        match value.as_sequence() {
+            None => {
+                // return Violation if value is not an Ion sequence
+                return Err(Violation::new(
+                    "contains",
+                    ViolationCode::TypeMismatched,
+                    &format!("expected list found {}", value.ion_type()),
+                ));
+            }
+            Some(ion_sequence) => {
+                // if any of our expected values are not in the Ion sequence then return Violation
+                if !expected_values.all(|ev| ion_sequence.iter().find(|v| v == &ev).is_some()) {
+                    return Err(Violation::new(
+                        "contains",
+                        ViolationCode::MissingValue,
+                        &format!("{:?} has missing value(s): {:?}", value, expected_values),
+                    ));
+                }
```

Comment:
Please correct me if I'm wrong, but I think the error message might be misleading. For example, given this type and value:
```
type::{ contains: [a, b, c] }
```
```
[ a, d, c ]
```
The validation error message would be:
```
[a, d, c] has missing value(s): [a, b, c] 
```
But it would be more accurate to say:
```
[a, d, c] has missing value(s): [b] 
```

Reply:
Yes, this is correct. I initially had  `expected_values` keep track of this but when I changed to use `.all()` method this was removed. Sorry for the confusion but I am changing back to a for loop over given values as if stick with `.all()` I would not be able to manipulate `expected_values` to keep track of missing values.